### PR TITLE
MMP-349 Lib folder generation for 2.x

### PR DESCRIPTION
**Pull request by WilliamPinho**

### CODE REVIEW

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/AbstractGenericMojo.java [hunk 1]
```diff
@@ -80,34 +80,26 @@
...
   protected String projectBuildDirectory;
 
-  // mule-packager
-
-  // @Parameter(readonly = true, required = true, defaultValue = "${project.remoteArtifactRepositories}")
-  // protected List<ArtifactRepository> remoteArtifactRepositories;
+  /**
+   * Name of the generated Mule App.
+   *
+   * @parameter alias="appName" default-value="${project.build.finalName}"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/AbstractGenericMojo.java]: this two annotations shouldn't be needed anymore

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java [hunk 2]
```diff
@@ -35,6 +40,53 @@
   @Component
   protected MavenProjectHelper helper;
 
+  /**
+   * Whether a JAR file will be created for the classes in the app. Using this optional configuration parameter will make the
+   * generated classes to be archived into a jar file and the classes directory will then be excluded from the app.
+   *
+   * @parameter expression="${archiveClasses}" default-value="false"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: this should be annotated with @parameter outside the comment

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java [hunk 3]
```diff
@@ -35,6 +40,53 @@
...
+   */
+  private boolean archiveClasses;
+
+
+  /**
+   * List of exclusion elements (having groupId and artifactId children) to exclude from the application archive.
+   *
+   * @parameter
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: outside the comment

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java [hunk 4]
```diff
@@ -35,6 +40,53 @@
...
+   * @since 2.3.0
+   */
+  private ArrayList<Exclusion> exclusions;
+
+  /**
+   * List of inclusion elements (having groupId and artifactId children) to exclude from the application archive.
+   *
+   * @parameter
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: outside the comment

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java [hunk 5]
```diff
@@ -35,6 +40,53 @@
...
+   * @since 2.3.0
+   */
+  private ArrayList<Inclusion> inclusions;
+
+  /**
+   * Exclude all artifacts with Mule groupIds. Default is <code>true</code>.
+   *
+   * @parameter default-value="true"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: outside the comment

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java [hunk 6]
```diff
@@ -35,6 +40,53 @@
...
+  private boolean excludeMuleDependencies;
+
+  /**
+   * @parameter default-value="false"
+   * @since 2.3.0
+   */
+  private boolean filterAppDirectory;
+
+  /**
+   * @parameter default-value="false"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: please all @parameter outside the comment section

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+   * @parameter
+   * @since 2.3.0
+   */
+  @Parameter
+  private List<Exclusion> exclusions;
+
+  /**
+   * List of inclusion elements (having groupId and artifactId children) to exclude from the application archive.
+   *
+   * @parameter
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java]: remove this

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+   * @since 2.3.0
+   */
+  @Parameter
+  private List<Inclusion> inclusions;
+
+  /**
+   * Exclude all artifacts with Mule groupIds. Default is <code>true</code>.
+   *
+   * @parameter default-value="true"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java]: remove this

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+   * @since 2.3.0
+   */
+  @Parameter(defaultValue = "true")
+  private boolean excludeMuleDependencies;
+
+  /**
+   * @parameter default-value="false"
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java]: remove this

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Exclusion.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.tools.maven.utils;
+
+public class Exclusion extends org.apache.maven.model.Exclusion implements org.mule.tools.api.packager.packaging.Exclusion {
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Exclusion.java]: Are there tests for this class? 
I don't see it in the commit, could we please add if missing?

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.maven.artifact.Artifact.SCOPE_COMPILE;
+import static org.apache.maven.artifact.Artifact.SCOPE_RUNTIME;
+
+public class Artifact implements org.mule.tools.api.util.Artifact {
```
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java]: Are there tests for this class? 
I don't see it in the commit, could we please add if missing?

File: mule-packager/src/main/java/org/mule/tools/api/packager/MuleProjectFoldersGenerator.java [hunk 12]
```diff
@@ -29,6 +29,8 @@ public void generate(Path targetFolder) {
 
     createFolderIfNecessary(targetFolder.toAbsolutePath().toString(), MULE.value());
 
+    createFolderIfNecessary(targetFolder.toAbsolutePath().toString(), LIB.value());
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/MuleProjectFoldersGenerator.java]: Can we please add a test for this class ?
We have one in 
https://github.com/mulesoft/mule-maven-plugin/blob/3.x/mule-packager/src/test/java/org/mule/tools/api/MuleProjectFoldersGeneratorTest.java

File: mule-packager/src/main/java/org/mule/tools/api/packager/archiver/MuleArchiver.java [hunk 13]
```diff
@@ -57,4 +62,14 @@ public void addWsdl(File resource, String[] includes, String[] excludes) throws
       addClasses(resource, includes, excludes);
     }
   }
+
+  /**
+   * @param resource Folder which contents are going to be added to the lib folder.
+   * @return
+   */
+  public void addLib(File resource, String[] includes, String[] excludes) {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/archiver/MuleArchiver.java]: I reckon the test I created in 3.x for this is not enough 
Could we please create a jira to update the test of this class and validate it properly ?

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java (hunk 11 above)
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java]: Is not a very good practice to name an implementation after its interface (though we do it some times ) 
Given the both are new and ours cane we rename this to DefaultArtifact or something like that?

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Exclusion.java (hunk 10 above)
Comment (damiansima) [inline mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Exclusion.java]: Is not a very good practice to name an implementation after its interface (though we do it some times ) 
Given the both are new and ours cane we rename this to DefaultArtifact or something like that?

File: mule-packager/src/main/java/org/mule/tools/api/packager/builder/PackageBuilderFactory.java [hunk 14]
```diff
@@ -6,15 +6,18 @@
...
 
+import org.mule.tools.api.packager.packaging.PackagingOptions;
+
 /**
  * Factory for package builders.
  */
 public class PackageBuilderFactory {
 
   /**
    * @return
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/builder/PackageBuilderFactory.java]: could we put a simple comment here 'returns a {@link PackageBuilder} based on the packaging options'

File: mule-packager/src/main/java/org/mule/tools/api/packager/filter/DependenciesFilter.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.util.*;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static java.util.Collections.unmodifiableSet;
+
+/**
+ * Filter to mule project dependencies.
+ */
+public class DependenciesFilter {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/filter/DependenciesFilter.java]: Could we please add a test for this class, it looks important we don't want it failing randomly 💃

File: mule-packager/src/main/java/org/mule/tools/api/packager/filter/prefix/PrefixTreeNode.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+package org.mule.tools.api.packager.filter.prefix;
+
+import java.util.Map;
+import java.util.TreeMap;
+
+class PrefixTreeNode {
+
+  private Map<Character, PrefixTreeNode> children = new TreeMap<>();
+
+  PrefixTreeNode addChild(Character c) {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/filter/prefix/PrefixTreeNode.java]: please add visibility to this methods public/protected/private

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class PackagingOptions {
+
+  /**
+   * Whether a JAR file will be created for the classes in the app. Using this optional configuration parameter will make the
+   * generated classes to be archived into a jar file and the classes directory will then be excluded from the app.
+   *
+   * @parameter expression="${archiveClasses}" default-value="false"
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  private final boolean archiveClasses;
+
+
+  /**
+   * List of exclusion elements (having groupId and artifactId children) to exclude from the application archive.
+   *
+   * @parameter
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+   * @since 3.2.0
+   */
+  private final ArrayList<Exclusion> exclusions;
+
+  /**
+   * List of inclusion elements (having groupId and artifactId children) to exclude from the application archive.
+   *
+   * @parameter
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+   * @since 3.2.0
+   */
+  private final ArrayList<Inclusion> inclusions;
+
+  /**
+   * Exclude all artifacts with Mule groupIds. Default is <code>true</code>.
+   *
+   * @parameter default-value="true"
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+   * @since 3.2.0
+   */
+  private final boolean excludeMuleDependencies;
+
+  /**
+   * @parameter default-value="false"
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+   * @since 3.2.0
+   */
+  private final boolean filterAppDirectory;
+
+  /**
+   * @parameter default-value="false"
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: Please remove or put it outside comment section

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/AbstractGenericMojo.java (hunk 1 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/AbstractGenericMojo.java]: Done!

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java (hunk 2 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: This parameter shouldn't be there so I deleted it

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java (hunk 3 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: Same here

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java (hunk 4 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: Idem

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java (hunk 5 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: Idem

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java (hunk 6 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/PackageMojo.java]: Idem

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java (hunk 7 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java]: Done

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java (hunk 9 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/mojo/ProcessResourcesMojo.java]: Done

File: mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java (hunk 11 above)
Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java]: done

Comment (WilliamPinho) [reply mule-maven-plugin/src/main/java/org/mule/tools/maven/utils/Artifact.java]: ok!

File: mule-packager/src/main/java/org/mule/tools/api/packager/MuleProjectFoldersGenerator.java (hunk 12 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/MuleProjectFoldersGenerator.java]: There is one, I added a test

File: mule-packager/src/main/java/org/mule/tools/api/packager/archiver/MuleArchiver.java (hunk 13 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/archiver/MuleArchiver.java]: Ok!

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 17 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: I deleted the class, so done

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 18 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: same

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 19 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: same

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 20 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: same

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 21 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: idem

File: mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java (hunk 22 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/packaging/PackagingOptions.java]: same

File: mule-packager/src/main/java/org/mule/tools/api/packager/filter/DependenciesFilter.java (hunk 15 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/packager/filter/DependenciesFilter.java]: Done!

**Pull request merged**